cpu/stm32/periph/usbdev_fs.c

```diff
@@ -0,0 +1,671 @@
+        if (*val & USB_EPTX_DTOG1) {
+            *val |= USB_EPTX_DTOG1;
+        }
+    }
+    if (mask & USB_EPTX_DTOG2) {
+        *val ^= USB_EPTX_DTOG2;
+    } else {
+        if (*val & USB_EPTX_DTOG2) {
+            *val |= USB_EPTX_DTOG2;
+        }
+    }
+}
+
+static void _set_ep_out_status(uint16_t *val, uint16_t mask)
+{
+       /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPRX_DTOG1) {
+        *val ^= USB_EPRX_DTOG1;
+    } else {
+        if (*val & USB_EPRX_DTOG1) {
+            *val |= USB_EPRX_DTOG1;
+        }
+    }
+    if (mask & USB_EPRX_DTOG2) {
+        *val ^= USB_EPRX_DTOG2;
+    } else {
+        if (*val & USB_EPRX_DTOG2) {
+            *val |= USB_EPRX_DTOG2;
+        }
+    }
+}
+
+static inline void _usb_attach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    bool syncing = false;
+    /* Enable DP pullup to signal connection */
+    _global_regs(conf)->BCDR |= USB_BCDR_DPPU;
+    do {
+        syncing = CRS->ISR & 0x00000008;
+    } while (!syncing);
+#else
+    (void)usbdev;
+#endif /* USB_BCDR_DPPU */
+}
+
+static inline void _usb_detach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    DEBUG_PUTS("usbdev_fs: Detaching from host");
+
+    /* Disable DP pullup to signal disconnection */
+    CLRBIT(_global_regs(conf)->BCDR, USB_BCDR_DPPU);
+#else
+    (void)usbdev;
+#endif
+}
+
+static void _set_address(stm32_usbdev_fs_t *usbdev, uint8_t address)
+{
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    _global_regs(conf)->DADDR = USB_DADDR_EF | (address & USB_DADDR_ADD);
+}
+
+void isr_usb_hp(void) {
+
+}
+
+void isr_usb_lp(void) {
+    stm32_usbdev_fs_t *usbdev = &_usbdevs[0];
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    int irq_status = _global_regs(conf)->ISTR;
+    /* Disable interrupts temporarily */
+    _global_regs(conf)->CNTR = 0;
+    /* Correct transfer Interrupt */
+    if (_global_regs(conf)->ISTR & USB_ISTR_CTR) {
+        /* Clear pending interrupts */
+        _global_regs(conf)->ISTR = 0;
+        /* Get endpoint number and associated ep pointer */
+        unsigned epnum = irq_status & 0x000F;
+        usbdev_ep_t *ep = (irq_status & USB_ISTR_DIR) ? &_ep_out[epnum] : &_ep_in[epnum];
+        /* get type and add endpoint address */
+        uint16_t reg = _type_to_reg(ep->type) | ep->num;
+
+        if (irq_status & USB_ISTR_DIR) {
+            /* Copy SRAM buffer to OUT buffer */
+            uint8_t* tmp_ptr = (uint8_t*)(USB1_PMAADDR + EP_DESC[ep->num].addr_rx);
+            for (uint8_t cpt=0; cpt<64; cpt++) {
+                _app_pbuf[ep->num][cpt] = tmp_ptr[cpt];
+            }
```

Comment:
Does it make sense to move this relative costly memory copy to outside the IRQ handling and into the `_usbdev_ep_esr` function?

Reply:
I'll give it a try.

Reply:
I'm fine with leaving this to a follow up, but would appreciate it if you can try it if you have time.

Reply:
Moved.

Reply:
This change doesn't seem to break anything so let's keep it.